Updated storage configuration to apply to events.

russjones opened the pull request:
**Description**

Updated storage configuration to not only apply to DynamoDB in the backend package, but also DynamoDB in the events package. This allows configuring continuous backups and auto scaling for the events table.

**Related Issues**

Fixes https://github.com/gravitational/teleport/pull/4662

r0mant commented on lib/events/dynamoevents/dynamoevents.go:
```diff
@@ -204,6 +237,28 @@ func New(cfg Config) (*Log, error) {
+
+	// Enable continuous backups if requested.
+	if b.Config.EnableContinuousBackups {
+		if err := dynamo.SetContinuousBackups(context.TODO(), b.svc, b.Tablename); err != nil {
```
I would consider passing context to the New() method and use here and below so these calls can be properly canceled.

r0mant commented on lib/events/dynamoevents/dynamoevents.go:
```diff
@@ -93,6 +109,19 @@ func (cfg *Config) CheckAndSetDefaults() error {
+
+	// If auto scaling is enabled, check all values are provided.
+	if cfg.EnableAutoScaling {
+		if cfg.ReadMaxCapacity == 0 || cfg.ReadMinCapacity == 0 || cfg.ReadTargetValue < 0.0 ||
```
I'm not that familiar with dynamodb, but isn't 0 a valid value for read/write min capacity?

russjones replied:
Removed validation, leaving it to AWS to return an error when invalid values are passed.

r0mant commented on lib/events/dynamoevents/dynamoevents.go:
```diff
@@ -60,6 +62,24 @@ type Config struct {
+	// EnableAutoScaling is used to enable (or disable) auto scaling policy.
+	EnableAutoScaling bool `json:"auto_scaling,omitempty"`
+	// ReadMaxCapacity is the maximum provisioned read capacity.
+	ReadMaxCapacity int
```
Any reason only two of the new fields have json tags set on them?

r0mant commented on lib/backend/dynamo/configure.go:
```diff
@@ -42,137 +41,80 @@ func (b *Backend) setContinuousBackups(ctx context.Context) error {
 }
 
-// disableAutoScaling disables auto scaling.
-func (b *Backend) disableAutoScaling(ctx context.Context) error {
```
So once enabled, there's no way to disable it anymore? Or is the "set" method above with zero values does equivalent thing?

russjones replied:
Good question.

In the initial implementation it was implemented if `auto_scaling: false` it would disable it. This had the unintended consequence that not setting it would cause Teleport to disable auto scaling. This isn't ideal because if someone is configuring DynamoDB elsewhere, Teleport could be disabling your auto scaling settings.

I could have made the bool a pointer but it just felt like it was getting too complex at that point.

a-palchikov commented on lib/backend/dynamo/configure_test.go:
```diff
@@ -21,28 +21,83 @@ package dynamo
+	// Check auto scaling values match.
+	resp, err := getAutoScaling(context.Background(), applicationautoscaling.New(d.session), d.Config.TableName)
+	require.NoError(t, err)
+	require.Equal(t, resp.ReadMinCapacity, 10)
```
Instead, use a struct-equivalency:
```go
require.Equal(t, resp, &AutoScalingParams{
	ReadMinCapacity:  10,
	ReadMinCapacity:  10,
	ReadMaxCapacity:  20,
	ReadTargetValue:  50.0,
	WriteMinCapacity: 10,
	WriteMaxCapacity: 20,
	WriteTargetValue: 50.0,
})
```
this will test unanticipated changes to the value.

a-palchikov commented on lib/backend/dynamo/configure_test.go:
```diff
@@ -21,28 +21,83 @@ package dynamo
+	require.Equal(t, resp.WriteTargetValue, 50.0)
+
+	// Remove table after tests are done.
+	err = deleteTable(context.Background(), d.svc, d.Config.TableName)
```
This is not re-entrancy friendly. If tests are executed multiple times and other tests would be influenced by this table, it is better to consistently cleanup using e.g. [testing.Cleanup](https://golang.org/pkg/testing/#T.Cleanup).

russjones replied:
Nice catch, fixed.

a-palchikov commented on lib/events/dynamoevents/dynamoevents.go:
```diff
@@ -60,6 +62,24 @@ type Config struct {
+	// EnableContinuousBackups is used to enable (or disable) PITR (Point-In-Time Recovery).
+	EnableContinuousBackups bool
+
+	// EnableAutoScaling is used to enable (or disable) auto scaling policy.
```
to enable (or disable) auto scaling.
Policy would define the type of (auto) scaling.

a-palchikov commented on lib/events/dynamoevents/dynamoevents.go:
```diff
@@ -93,6 +109,12 @@ func (cfg *Config) CheckAndSetDefaults() error {
 		cfg.UIDGenerator = utils.NewRealUID()
 	}
+
+	// Table is required.
```
Not sure why you'd move the test after the defaults - usually defaults are applied to a valid (empty) state, so it looked to have been fine before?

a-palchikov commented on lib/events/dynamoevents/dynamoevents.go:
```diff
@@ -204,6 +230,28 @@ func New(cfg Config) (*Log, error) {
 	}
+
+	// Enable continuous backups if requested.
+	if b.Config.EnableContinuousBackups {
```
How can continuous backups and auto-scaling be disabled? Or teleport does not make an attempt to do that?
I'm asking since the switches were advertised as working both ways but seems like they're used only to turn settings on. I also read the previous comment about the choice of bool vs pointer to bool which would've solved it but it was explicitly chosen to favor simplicity which means just turning it on?

russjones replied:
Yeah, now it only supports turning on and configuring.

a-palchikov commented on lib/backend/dynamo/configure.go:
```diff
@@ -42,137 +41,80 @@ func (b *Backend) setContinuousBackups(ctx context.Context) error {
-		return b.enableAutoScaling(ctx)
-	}
-	return b.disableAutoScaling(ctx)
+// AutoScalingParams define auto scaling parameters for DynamoDB.
```
nit // AutoScalingParams define**s** auto scaling parameters for DynamoDB.

pierrebeaucamp commented on lib/service/service.go:
```diff
@@ -923,15 +923,23 @@ func initExternalLog(auditConfig services.AuditConfig) (events.IAuditLog, error)
+				ReadMinCapacity:         int(auditConfig.ReadMinCapacity),
+				ReadTargetValue:         auditConfig.ReadTargetValue,
+				WriteMaxCapacity:        int(auditConfig.WriteMaxCapacity),
+				WriteMinCapacity:        int(auditConfig.WriteMaxCapacity),
```
There is a subtle error here: You're setting the minimum capacity to the maximum capacity value.

pierrebeaucamp replied:
Other than this, the branch seems to be working fine 👍

russjones commented on lib/backend/dynamo/configure_test.go:
```diff
@@ -21,28 +21,83 @@ package dynamo
+func TestAutoScaling(t *testing.T) {
+	// Create new backend with auto scaling enabled.
+	newBackend := func() (backend.Backend, error) {
+		return New(context.Background(), map[string]interface{}{
```
We're moving away from this and to a `Config` type, for example that's how `lib/events/dynamoevents` does it. Didn't want to do that refactoring during this PR.

russjones committed:
Updated storage configuration to apply to events.

Updated storage configuration to not only apply to DynamoDB in the
backend package, but also DynamoDB in the events package. This allows
configuring continuous backups and auto scaling for the events table.

russjones commented:
@pierrebeaucamp @a-palchikov Can you take another look?

pierrebeaucamp commented:
@benarent as discussed yesterday: for autoscaling to work, we need to add the following IAM permissions:

```
        {
            "Action": [
                "application-autoscaling:PutScalingPolicy",
                "application-autoscaling:RegisterScalableTarget"
            ],
            "Effect": "Allow",
            "Resource": "*"
        },
        {
            "Action": [
                "iam:CreateServiceLinkedRole"
            ],
            "Condition": {
                "StringEquals": {
                    "iam:AWSServiceName": [
                        "dynamodb.application-autoscaling.amazonaws.com"
                    ]
                }
            },
            "Effect": "Allow",
            "Resource": "*"
        }
```

r0mant commented on lib/service/service.go:
```diff
@@ -923,15 +923,23 @@ func initExternalLog(auditConfig services.AuditConfig) (events.IAuditLog, error)
+				Region:                  auditConfig.Region,
+				EnableContinuousBackups: auditConfig.EnableContinuousBackups,
+				EnableAutoScaling:       auditConfig.EnableAutoScaling,
+				ReadMinCapacity:         int(auditConfig.ReadMinCapacity),
```
Just curious, would it make sense to make `ReadMinCapacity` and other new fields be of a consistent type everywhere - either `int` or `int64` and avoid conversions (and potentially int overflow on 32 bit systems)? Seems like it's `int64` in the proto file but `int` in the config struct. I would prob. go with the type that AWS Go SDK accepts. Disregard if there's a technical reason for this, or too much refactoring required.

r0mant commented on lib/events/dynamoevents/dynamoevents.go:
```diff
@@ -60,6 +62,24 @@ type Config struct {
+	// ReadMaxCapacity is the maximum provisioned read capacity.
+	ReadMaxCapacity int
+	// ReadMinCapacity is the minimum provisioned read capacity.
+	ReadMinCapacity int
```
I'm still kinda curious why these new field don't have json tags on them - doesn't this struct basically map onto the file config (like its godoc suggests), so how do they look in the config?

IIRC the default behavior is that for json they marshal into the same name as the Go field name (so `ReadMaxCapacity: 1000`, etc.) if the json tag is not set.

russjones replied:
`AuditConfig` is the type that's read from file configuration, it's defined in types.proto:

https://github.com/gravitational/teleport/blob/master/lib/services/types.proto#L461-L480

That's then used to set `dynamoevents.Config`. The old json tags approach was probably just migrated away from and they were never removed. Which is why I didn't include them in any new code.

https://github.com/gravitational/teleport/blob/rjones/aws-fixes/lib/service/service.go#L923-L936

russjones committed:
Merge branch 'master' into rjones/aws-fixes
